# My Experiment - PIMA Indians Diabetes Prediction with Deep Learning

While the previous notebook in the folder was about hyperparamter tuning by grid-search for the prediction of PIMA Indians Diabetes, this notebook tries to focus on the weaknesses, which were observed in the previous notebook. Fact is:
* There was no train test split, so it was not easy to judge how good the model was, when it will be applied on any unseen data.
* The dataset was imbalanced, leading to the result that the prediction for the minority class (1) was less good than the majority class (0).

In [1]:
import sys
import pandas as pd
import numpy as np
import sklearn
import keras
import tensorflow as tf

Using TensorFlow backend.


In [2]:
seed = 6
np.random.seed(seed)
tf.random.set_seed(seed)

In [3]:
fileName = "diabetes.csv"
columns = ['n_pregnant', 'glucose_concentration', 'blood_pressure (mm Hg)', 'skin_thickness (mm)', 'serum_insulin (mu U/ml)',
        'BMI', 'pedigree_function', 'age', 'class']
df = pd.read_csv(fileName, names = columns, header=0)

In [4]:
df

n_pregnant  glucose_concentration  blood_pressure (mm Hg)  \
0             6                    148                      72   
1             1                     85                      66   
2             8                    183                      64   
3             1                     89                      66   
4             0                    137                      40   
..          ...                    ...                     ...   
763          10                    101                      76   
764           2                    122                      70   
765           5                    121                      72   
766           1                    126                      60   
767           1                     93                      70   

     skin_thickness (mm)  serum_insulin (mu U/ml)   BMI  pedigree_function  \
0                     35                        0  33.6              0.627   
1                     29                        0  26.6              0.351   
2                      0                        0  23.3              0.672   
3                     23                       94  28.1              0.167   
4                     35                      168  43.1              2.288   
..                   ...                      ...   ...                ...   
763                   48                      180  32.9              0.171   
764                   27                        0  36.8              0.340   
765                   23                      112  26.2              0.245   
766                    0                        0  30.1              0.349   
767                   31                        0  30.4              0.315   

     age  class  
0     50      1  
1     31      0  
2     32      1  
3     21      0  
4     33      1  
..   ...    ...  
763   63      0  
764   27      0  
765   30      0  
766   47      1  
767   23      0  

[768 rows x 9 columns]

In [5]:
columns = ['glucose_concentration', 'blood_pressure (mm Hg)', 'skin_thickness (mm)', 'serum_insulin (mu U/ml)', 'BMI']

for col in columns:
    df[col].replace(0, np.NaN, inplace=True)
    
df.dropna(inplace=True)
df

n_pregnant  glucose_concentration  blood_pressure (mm Hg)  \
3             1                   89.0                    66.0   
4             0                  137.0                    40.0   
6             3                   78.0                    50.0   
8             2                  197.0                    70.0   
13            1                  189.0                    60.0   
..          ...                    ...                     ...   
753           0                  181.0                    88.0   
755           1                  128.0                    88.0   
760           2                   88.0                    58.0   
763          10                  101.0                    76.0   
765           5                  121.0                    72.0   

     skin_thickness (mm)  serum_insulin (mu U/ml)   BMI  pedigree_function  \
3                   23.0                     94.0  28.1              0.167   
4                   35.0                    168.0  43.1              2.288   
6                   32.0                     88.0  31.0              0.248   
8                   45.0                    543.0  30.5              0.158   
13                  23.0                    846.0  30.1              0.398   
..                   ...                      ...   ...                ...   
753                 44.0                    510.0  43.3              0.222   
755                 39.0                    110.0  36.5              1.057   
760                 26.0                     16.0  28.4              0.766   
763                 48.0                    180.0  32.9              0.171   
765                 23.0                    112.0  26.2              0.245   

     age  class  
3     21      0  
4     33      1  
6     26      1  
8     53      1  
13    59      1  
..   ...    ...  
753   26      1  
755   37      1  
760   22      0  
763   63      0  
765   30      0  

[392 rows x 9 columns]

In [8]:
dataset = df.values
print(dataset.shape)

X = dataset[:,0:8]
y = dataset[:, 8].astype(int)

print(X.shape)
print(y.shape)

(392, 9)
(392, 8)
(392,)


The previous steps are pretty similar compared to the first notebook. We dropped the NaN values, 393 rows are remaining. Having `X` and `y`: We will split the data into train and test sets, which means the amount of training material will be less than 392, but we try to not cause underfitting.

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=seed, shuffle=True)

print(X_train.shape)
print(X_test.shape)

(313, 8)
(79, 8)


In [10]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler().fit(X_train)
print(scaler)

X_standardized = scaler.transform(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)


In [11]:
from sklearn.model_selection import GridSearchCV, KFold
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.optimizers import Adam

The previous notebook showed which parameters were working well in each of the experimented combinations. I will apply some findings in this notebook.

In [12]:
def create_model():
    model = Sequential()
    model.add(Dense(4, input_dim = 8, kernel_initializer='normal', activation='linear'))
    model.add(Dense(2, input_dim = 4, kernel_initializer='normal', activation='linear'))
    model.add(Dense(1, activation='sigmoid'))
    
    adam = Adam(lr = 0.01)
    model.compile(loss = 'binary_crossentropy', optimizer = adam, metrics = ['accuracy'])
    return model

model = create_model()
print(model.summary())

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 4)                 36        
_________________________________________________________________
dense_2 (Dense)              (None, 2)                 10        
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 3         
Total params: 49
Trainable params: 49
Non-trainable params: 0
_________________________________________________________________
None


Again, as the goal of this week is experiment with grid-search, we will continue the experiment trying to find the best batch size and number of epochs. The difference is, this time I added the paramter class weight to the fit function. Class 1 is the minority class, that is why I assigned a greater weight to it than to the majority class 0.

In [13]:
# create the model
model = KerasClassifier(build_fn = create_model, verbose = 1)

# define the grid search parameters
batch_size = [10, 20, 40]
epochs = [10, 50, 100]

# make a dictionary of the grid search parameters
param_grid = dict(batch_size=batch_size, epochs=epochs)

# build and fit the GridSearchCV
grid = GridSearchCV(estimator = model, param_grid = param_grid, cv = KFold(shuffle=True, random_state=seed), refit=True, verbose=20, return_train_score=True)
grid_results = grid.fit(X_standardized, y_train, class_weight={0: 1.0, 1: 2.0})

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV] batch_size=10, epochs=10 ........................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Epoch 1/10
250/250 [==============================] - 0s 1ms/step - loss: 0.8385 - accuracy: 0.5760
Epoch 2/10
250/250 [==============================] - 0s 182us/step - loss: 0.6650 - accuracy: 0.7400
Epoch 3/10
250/250 [==============================] - 0s 162us/step - loss: 0.6494 - accuracy: 0.7760
Epoch 4/10
250/250 [==============================] - 0s 171us/step - loss: 0.6331 - accuracy: 0.7840
Epoch 5/10
250/250 [==============================] - 0s 289us/step - loss: 0.6396 - accuracy: 0.7760
Epoch 6/10
250/250 [==============================] - 0s 952us/step - loss: 0.6353 - accuracy: 0.7680
Epoch 7/10
250/250 [==============================] - 0s 516us/step - loss: 0.6329 - accuracy: 0.7880
Epoch 8/10
250/250 [==============================] - 0s 707us/step - loss: 0.6329 - accuracy: 0.7880
Epoch 9/10
250/250 [==============================] - 0s 513us/step - loss: 0.6478 - accuracy: 0.7560
Epoch 10/10
250/250 [==============================] - 0s 105us/step
[CV]  batch_siz

[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    2.2s remaining:    0.0s


Epoch 1/10
250/250 [==============================] - 0s 746us/step - loss: 0.8524 - accuracy: 0.6960
Epoch 2/10
250/250 [==============================] - 0s 176us/step - loss: 0.6711 - accuracy: 0.7560
Epoch 3/10
250/250 [==============================] - 0s 153us/step - loss: 0.6484 - accuracy: 0.7880
Epoch 4/10
250/250 [==============================] - 0s 164us/step - loss: 0.6386 - accuracy: 0.7960
Epoch 5/10
250/250 [==============================] - 0s 148us/step - loss: 0.6356 - accuracy: 0.7960
Epoch 6/10
250/250 [==============================] - 0s 205us/step - loss: 0.6317 - accuracy: 0.8040
Epoch 7/10
250/250 [==============================] - 0s 149us/step - loss: 0.6287 - accuracy: 0.8040
Epoch 8/10
250/250 [==============================] - 0s 175us/step - loss: 0.6319 - accuracy: 0.8080
Epoch 9/10
250/250 [==============================] - 0s 154us/step - loss: 0.6277 - accuracy: 0.8080
Epoch 10/10
250/250 [==============================] - 0s 71us/step
[CV]  batch_si

[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    3.7s remaining:    0.0s


Epoch 1/10
250/250 [==============================] - 0s 737us/step - loss: 0.8533 - accuracy: 0.6640
Epoch 2/10
250/250 [==============================] - 0s 152us/step - loss: 0.7205 - accuracy: 0.7320
Epoch 3/10
250/250 [==============================] - 0s 192us/step - loss: 0.6912 - accuracy: 0.7640
Epoch 4/10
250/250 [==============================] - 0s 160us/step - loss: 0.6908 - accuracy: 0.7560
Epoch 5/10
250/250 [==============================] - 0s 145us/step - loss: 0.6886 - accuracy: 0.7640
Epoch 6/10
250/250 [==============================] - 0s 151us/step - loss: 0.6817 - accuracy: 0.7560
Epoch 7/10
250/250 [==============================] - 0s 155us/step - loss: 0.6838 - accuracy: 0.7640
Epoch 8/10
250/250 [==============================] - 0s 161us/step - loss: 0.6759 - accuracy: 0.7640
Epoch 9/10
250/250 [==============================] - 0s 182us/step - loss: 0.6805 - accuracy: 0.7640
Epoch 10/10
250/250 [==============================] - 0s 67us/step
[CV]  batch_si

[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    5.0s remaining:    0.0s


Epoch 1/10
251/251 [==============================] - 0s 768us/step - loss: 0.8329 - accuracy: 0.6853
Epoch 2/10
251/251 [==============================] - 0s 155us/step - loss: 0.7084 - accuracy: 0.7331
Epoch 3/10
251/251 [==============================] - 0s 177us/step - loss: 0.6935 - accuracy: 0.7450
Epoch 4/10
251/251 [==============================] - 0s 150us/step - loss: 0.6818 - accuracy: 0.7729
Epoch 5/10
251/251 [==============================] - 0s 173us/step - loss: 0.6832 - accuracy: 0.7610
Epoch 6/10
251/251 [==============================] - 0s 175us/step - loss: 0.6749 - accuracy: 0.7610
Epoch 7/10
251/251 [==============================] - 0s 159us/step - loss: 0.6880 - accuracy: 0.7689
Epoch 8/10
251/251 [==============================] - 0s 149us/step - loss: 0.6826 - accuracy: 0.7610
Epoch 9/10
251/251 [==============================] - 0s 144us/step - loss: 0.6786 - accuracy: 0.7649
Epoch 10/10
251/251 [==============================] - 0s 70us/step
[CV]  batch_si

[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    6.2s remaining:    0.0s


Epoch 1/10
251/251 [==============================] - 0s 787us/step - loss: 0.8332 - accuracy: 0.6096
Epoch 2/10
251/251 [==============================] - 0s 158us/step - loss: 0.6657 - accuracy: 0.7331
Epoch 3/10
251/251 [==============================] - 0s 150us/step - loss: 0.6509 - accuracy: 0.7490
Epoch 4/10
251/251 [==============================] - 0s 170us/step - loss: 0.6519 - accuracy: 0.7610
Epoch 5/10
251/251 [==============================] - 0s 155us/step - loss: 0.6415 - accuracy: 0.7649
Epoch 6/10
251/251 [==============================] - 0s 148us/step - loss: 0.6416 - accuracy: 0.7649
Epoch 7/10
251/251 [==============================] - 0s 177us/step - loss: 0.6371 - accuracy: 0.7729
Epoch 8/10
251/251 [==============================] - 0s 202us/step - loss: 0.6474 - accuracy: 0.7689
Epoch 9/10
251/251 [==============================] - 0s 171us/step - loss: 0.6367 - accuracy: 0.7649
Epoch 10/10
251/251 [==============================] - 0s 71us/step
[CV]  batch_si

[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    7.5s remaining:    0.0s


Epoch 1/50
250/250 [==============================] - 0s 717us/step - loss: 0.8408 - accuracy: 0.6720
Epoch 2/50
250/250 [==============================] - 0s 145us/step - loss: 0.6824 - accuracy: 0.7160
Epoch 3/50
250/250 [==============================] - 0s 142us/step - loss: 0.6439 - accuracy: 0.7520
Epoch 4/50
250/250 [==============================] - 0s 167us/step - loss: 0.6350 - accuracy: 0.7720
Epoch 5/50
250/250 [==============================] - 0s 145us/step - loss: 0.6384 - accuracy: 0.7640
Epoch 6/50
250/250 [==============================] - 0s 149us/step - loss: 0.6388 - accuracy: 0.7800
Epoch 7/50
250/250 [==============================] - 0s 176us/step - loss: 0.6373 - accuracy: 0.7920
Epoch 8/50
250/250 [==============================] - 0s 140us/step - loss: 0.6343 - accuracy: 0.7760
Epoch 9/50
250/250 [==============================] - 0s 146us/step - loss: 0.6366 - accuracy: 0.7600
Epoch 10/50
250/250 [==============================] - 0s 169us/step - loss: 0.633

[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:   11.2s remaining:    0.0s


Epoch 1/50
250/250 [==============================] - 0s 706us/step - loss: 0.8289 - accuracy: 0.6080
Epoch 2/50
250/250 [==============================] - 0s 143us/step - loss: 0.6577 - accuracy: 0.7800
Epoch 3/50
250/250 [==============================] - 0s 150us/step - loss: 0.6375 - accuracy: 0.7920
Epoch 4/50
250/250 [==============================] - 0s 144us/step - loss: 0.6376 - accuracy: 0.7880
Epoch 5/50
250/250 [==============================] - 0s 153us/step - loss: 0.6371 - accuracy: 0.8000
Epoch 6/50
250/250 [==============================] - 0s 151us/step - loss: 0.6330 - accuracy: 0.8000
Epoch 7/50
250/250 [==============================] - 0s 150us/step - loss: 0.6355 - accuracy: 0.7960
Epoch 8/50
250/250 [==============================] - 0s 139us/step - loss: 0.6290 - accuracy: 0.8040
Epoch 9/50
250/250 [==============================] - 0s 136us/step - loss: 0.6336 - accuracy: 0.7960
Epoch 10/50
250/250 [==============================] - 0s 141us/step - loss: 0.629

[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:   13.9s remaining:    0.0s


Epoch 1/50
250/250 [==============================] - 0s 717us/step - loss: 0.8704 - accuracy: 0.6000
Epoch 2/50
250/250 [==============================] - 0s 145us/step - loss: 0.7237 - accuracy: 0.7400
Epoch 3/50
250/250 [==============================] - 0s 218us/step - loss: 0.6974 - accuracy: 0.7560
Epoch 4/50
250/250 [==============================] - 0s 258us/step - loss: 0.6854 - accuracy: 0.7480
Epoch 5/50
250/250 [==============================] - 0s 204us/step - loss: 0.6831 - accuracy: 0.7600
Epoch 6/50
250/250 [==============================] - 0s 182us/step - loss: 0.6848 - accuracy: 0.7560
Epoch 7/50
250/250 [==============================] - 0s 154us/step - loss: 0.6816 - accuracy: 0.7480
Epoch 8/50
250/250 [==============================] - 0s 153us/step - loss: 0.6809 - accuracy: 0.7560
Epoch 9/50
250/250 [==============================] - 0s 139us/step - loss: 0.6848 - accuracy: 0.7600
Epoch 10/50
250/250 [==============================] - 0s 134us/step - loss: 0.676

[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:   16.9s remaining:    0.0s


Epoch 1/50
251/251 [==============================] - 0s 824us/step - loss: 0.8135 - accuracy: 0.6892
Epoch 2/50
251/251 [==============================] - 0s 148us/step - loss: 0.6945 - accuracy: 0.7410
Epoch 3/50
251/251 [==============================] - 0s 155us/step - loss: 0.6873 - accuracy: 0.7729
Epoch 4/50
251/251 [==============================] - 0s 158us/step - loss: 0.6844 - accuracy: 0.7729
Epoch 5/50
251/251 [==============================] - 0s 197us/step - loss: 0.6807 - accuracy: 0.7689
Epoch 6/50
251/251 [==============================] - 0s 200us/step - loss: 0.6784 - accuracy: 0.7689
Epoch 7/50
251/251 [==============================] - 0s 213us/step - loss: 0.6761 - accuracy: 0.7610
Epoch 8/50
251/251 [==============================] - 0s 246us/step - loss: 0.6701 - accuracy: 0.7689
Epoch 9/50
251/251 [==============================] - 0s 165us/step - loss: 0.6720 - accuracy: 0.7729
Epoch 10/50
251/251 [==============================] - 0s 150us/step - loss: 0.686

[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:   19.9s remaining:    0.0s


Epoch 1/50
251/251 [==============================] - 0s 713us/step - loss: 0.7878 - accuracy: 0.6773
Epoch 2/50
251/251 [==============================] - 0s 146us/step - loss: 0.6588 - accuracy: 0.7610
Epoch 3/50
251/251 [==============================] - 0s 151us/step - loss: 0.6528 - accuracy: 0.7570
Epoch 4/50
251/251 [==============================] - 0s 149us/step - loss: 0.6410 - accuracy: 0.7610
Epoch 5/50
251/251 [==============================] - 0s 154us/step - loss: 0.6479 - accuracy: 0.7570
Epoch 6/50
251/251 [==============================] - 0s 144us/step - loss: 0.6393 - accuracy: 0.7689
Epoch 7/50
251/251 [==============================] - 0s 142us/step - loss: 0.6398 - accuracy: 0.7849
Epoch 8/50
251/251 [==============================] - 0s 142us/step - loss: 0.6408 - accuracy: 0.7649
Epoch 9/50
251/251 [==============================] - 0s 143us/step - loss: 0.6364 - accuracy: 0.7729
Epoch 10/50
251/251 [==============================] - 0s 143us/step - loss: 0.640

[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:   22.6s remaining:    0.0s


Epoch 1/100
250/250 [==============================] - 0s 758us/step - loss: 0.8239 - accuracy: 0.7360
Epoch 2/100
250/250 [==============================] - 0s 144us/step - loss: 0.6688 - accuracy: 0.7480
Epoch 3/100
250/250 [==============================] - 0s 139us/step - loss: 0.6473 - accuracy: 0.7520
Epoch 4/100
250/250 [==============================] - 0s 142us/step - loss: 0.6416 - accuracy: 0.7760
Epoch 5/100
250/250 [==============================] - 0s 144us/step - loss: 0.6371 - accuracy: 0.7840
Epoch 6/100
250/250 [==============================] - 0s 140us/step - loss: 0.6381 - accuracy: 0.7800
Epoch 7/100
250/250 [==============================] - 0s 137us/step - loss: 0.6396 - accuracy: 0.7720
Epoch 8/100
250/250 [==============================] - 0s 142us/step - loss: 0.6345 - accuracy: 0.7760
Epoch 9/100
250/250 [==============================] - 0s 138us/step - loss: 0.6330 - accuracy: 0.7680
Epoch 10/100
250/250 [==============================] - 0s 175us/step - l

250/250 [==============================] - 0s 140us/step - loss: 0.6273 - accuracy: 0.7840
Epoch 80/100
250/250 [==============================] - 0s 150us/step - loss: 0.6281 - accuracy: 0.7840
Epoch 81/100
250/250 [==============================] - 0s 136us/step - loss: 0.6327 - accuracy: 0.7600
Epoch 82/100
250/250 [==============================] - 0s 138us/step - loss: 0.6360 - accuracy: 0.7800
Epoch 83/100
250/250 [==============================] - 0s 137us/step - loss: 0.6283 - accuracy: 0.7760
Epoch 84/100
250/250 [==============================] - 0s 136us/step - loss: 0.6295 - accuracy: 0.7880
Epoch 85/100
250/250 [==============================] - 0s 137us/step - loss: 0.6286 - accuracy: 0.7680
Epoch 86/100
250/250 [==============================] - 0s 137us/step - loss: 0.6301 - accuracy: 0.7840
Epoch 87/100
250/250 [==============================] - 0s 137us/step - loss: 0.6284 - accuracy: 0.7800
Epoch 88/100
250/250 [==============================] - 0s 139us/step - loss:

[Parallel(n_jobs=1)]: Done  11 out of  11 | elapsed:   27.7s remaining:    0.0s


Epoch 1/100
250/250 [==============================] - 0s 802us/step - loss: 0.8044 - accuracy: 0.7520
Epoch 2/100
250/250 [==============================] - 0s 193us/step - loss: 0.6758 - accuracy: 0.7600
Epoch 3/100
250/250 [==============================] - 0s 167us/step - loss: 0.6395 - accuracy: 0.7920
Epoch 4/100
250/250 [==============================] - 0s 173us/step - loss: 0.6337 - accuracy: 0.7920
Epoch 5/100
250/250 [==============================] - 0s 169us/step - loss: 0.6309 - accuracy: 0.7920
Epoch 6/100
250/250 [==============================] - 0s 163us/step - loss: 0.6350 - accuracy: 0.8000
Epoch 7/100
250/250 [==============================] - 0s 163us/step - loss: 0.6375 - accuracy: 0.8040
Epoch 8/100
250/250 [==============================] - 0s 162us/step - loss: 0.6329 - accuracy: 0.8080
Epoch 9/100
250/250 [==============================] - 0s 163us/step - loss: 0.6358 - accuracy: 0.8040
Epoch 10/100
250/250 [==============================] - 0s 160us/step - l

250/250 [==============================] - 0s 187us/step - loss: 0.6317 - accuracy: 0.8080
Epoch 80/100
250/250 [==============================] - 0s 178us/step - loss: 0.6258 - accuracy: 0.8120
Epoch 81/100
250/250 [==============================] - 0s 174us/step - loss: 0.6391 - accuracy: 0.8040
Epoch 82/100
250/250 [==============================] - 0s 182us/step - loss: 0.6334 - accuracy: 0.8080
Epoch 83/100
250/250 [==============================] - 0s 533us/step - loss: 0.6324 - accuracy: 0.8120
Epoch 84/100
250/250 [==============================] - 0s 296us/step - loss: 0.6261 - accuracy: 0.8080
Epoch 85/100
250/250 [==============================] - 0s 320us/step - loss: 0.6261 - accuracy: 0.8040
Epoch 86/100
250/250 [==============================] - 0s 296us/step - loss: 0.6280 - accuracy: 0.8080
Epoch 87/100
250/250 [==============================] - 0s 324us/step - loss: 0.6315 - accuracy: 0.8080
Epoch 88/100
250/250 [==============================] - 0s 207us/step - loss:

[Parallel(n_jobs=1)]: Done  12 out of  12 | elapsed:   35.2s remaining:    0.0s


Epoch 1/100
250/250 [==============================] - 0s 1ms/step - loss: 0.8560 - accuracy: 0.6040
Epoch 2/100
250/250 [==============================] - 0s 201us/step - loss: 0.7171 - accuracy: 0.7480
Epoch 3/100
250/250 [==============================] - 0s 207us/step - loss: 0.6907 - accuracy: 0.7480
Epoch 4/100
250/250 [==============================] - 0s 218us/step - loss: 0.6927 - accuracy: 0.7640
Epoch 5/100
250/250 [==============================] - 0s 173us/step - loss: 0.6873 - accuracy: 0.7440
Epoch 6/100
250/250 [==============================] - 0s 163us/step - loss: 0.6833 - accuracy: 0.7480
Epoch 7/100
250/250 [==============================] - 0s 169us/step - loss: 0.6798 - accuracy: 0.7680
Epoch 8/100
250/250 [==============================] - 0s 162us/step - loss: 0.6858 - accuracy: 0.7600
Epoch 9/100
250/250 [==============================] - 0s 239us/step - loss: 0.6775 - accuracy: 0.7680
Epoch 10/100
250/250 [==============================] - 0s 215us/step - los

250/250 [==============================] - 0s 140us/step - loss: 0.6776 - accuracy: 0.7640
Epoch 80/100
250/250 [==============================] - 0s 148us/step - loss: 0.6792 - accuracy: 0.7680
Epoch 81/100
250/250 [==============================] - 0s 146us/step - loss: 0.6761 - accuracy: 0.7680
Epoch 82/100
250/250 [==============================] - 0s 136us/step - loss: 0.6768 - accuracy: 0.7600
Epoch 83/100
250/250 [==============================] - 0s 159us/step - loss: 0.6821 - accuracy: 0.7600
Epoch 84/100
250/250 [==============================] - 0s 155us/step - loss: 0.6784 - accuracy: 0.7640
Epoch 85/100
250/250 [==============================] - 0s 170us/step - loss: 0.6766 - accuracy: 0.7680
Epoch 86/100
250/250 [==============================] - 0s 165us/step - loss: 0.6783 - accuracy: 0.7640
Epoch 87/100
250/250 [==============================] - 0s 160us/step - loss: 0.6846 - accuracy: 0.7600
Epoch 88/100
250/250 [==============================] - 0s 154us/step - loss:

[Parallel(n_jobs=1)]: Done  13 out of  13 | elapsed:   41.1s remaining:    0.0s


Epoch 1/100
251/251 [==============================] - 0s 818us/step - loss: 0.8288 - accuracy: 0.5777
Epoch 2/100
251/251 [==============================] - 0s 148us/step - loss: 0.7008 - accuracy: 0.7291
Epoch 3/100
251/251 [==============================] - 0s 152us/step - loss: 0.6893 - accuracy: 0.7331
Epoch 4/100
251/251 [==============================] - 0s 147us/step - loss: 0.6730 - accuracy: 0.7769
Epoch 5/100
251/251 [==============================] - 0s 144us/step - loss: 0.6825 - accuracy: 0.7769
Epoch 6/100
251/251 [==============================] - 0s 145us/step - loss: 0.6773 - accuracy: 0.7769
Epoch 7/100
251/251 [==============================] - 0s 144us/step - loss: 0.6679 - accuracy: 0.7689
Epoch 8/100
251/251 [==============================] - 0s 147us/step - loss: 0.6746 - accuracy: 0.7729
Epoch 9/100
251/251 [==============================] - 0s 155us/step - loss: 0.6712 - accuracy: 0.7769
Epoch 10/100
251/251 [==============================] - 0s 164us/step - l

251/251 [==============================] - 0s 165us/step - loss: 0.6674 - accuracy: 0.7610
Epoch 80/100
251/251 [==============================] - 0s 168us/step - loss: 0.6699 - accuracy: 0.7649
Epoch 81/100
251/251 [==============================] - 0s 146us/step - loss: 0.6724 - accuracy: 0.7649
Epoch 82/100
251/251 [==============================] - 0s 164us/step - loss: 0.6710 - accuracy: 0.7769
Epoch 83/100
251/251 [==============================] - 0s 200us/step - loss: 0.6720 - accuracy: 0.7769
Epoch 84/100
251/251 [==============================] - 0s 186us/step - loss: 0.6724 - accuracy: 0.7729
Epoch 85/100
251/251 [==============================] - 0s 185us/step - loss: 0.6699 - accuracy: 0.7729
Epoch 86/100
251/251 [==============================] - 0s 177us/step - loss: 0.6684 - accuracy: 0.7729
Epoch 87/100
251/251 [==============================] - 0s 154us/step - loss: 0.6698 - accuracy: 0.7888
Epoch 88/100
251/251 [==============================] - 0s 143us/step - loss:

[Parallel(n_jobs=1)]: Done  14 out of  14 | elapsed:   46.8s remaining:    0.0s


Epoch 1/100
251/251 [==============================] - 0s 715us/step - loss: 0.8544 - accuracy: 0.6454
Epoch 2/100
251/251 [==============================] - 0s 147us/step - loss: 0.6803 - accuracy: 0.7490
Epoch 3/100
251/251 [==============================] - 0s 145us/step - loss: 0.6726 - accuracy: 0.7610
Epoch 4/100
251/251 [==============================] - 0s 147us/step - loss: 0.6505 - accuracy: 0.7610
Epoch 5/100
251/251 [==============================] - 0s 144us/step - loss: 0.6475 - accuracy: 0.7689
Epoch 6/100
251/251 [==============================] - 0s 144us/step - loss: 0.6431 - accuracy: 0.7530
Epoch 7/100
251/251 [==============================] - 0s 148us/step - loss: 0.6365 - accuracy: 0.7689
Epoch 8/100
251/251 [==============================] - 0s 147us/step - loss: 0.6398 - accuracy: 0.7649
Epoch 9/100
251/251 [==============================] - 0s 150us/step - loss: 0.6368 - accuracy: 0.7689
Epoch 10/100
251/251 [==============================] - 0s 151us/step - l

251/251 [==============================] - 0s 143us/step - loss: 0.6350 - accuracy: 0.7649
Epoch 80/100
251/251 [==============================] - 0s 144us/step - loss: 0.6318 - accuracy: 0.7689
Epoch 81/100
251/251 [==============================] - 0s 137us/step - loss: 0.6355 - accuracy: 0.7570
Epoch 82/100
251/251 [==============================] - 0s 140us/step - loss: 0.6300 - accuracy: 0.7570
Epoch 83/100
251/251 [==============================] - 0s 139us/step - loss: 0.6327 - accuracy: 0.7649
Epoch 84/100
251/251 [==============================] - 0s 141us/step - loss: 0.6375 - accuracy: 0.7769
Epoch 85/100
251/251 [==============================] - 0s 143us/step - loss: 0.6305 - accuracy: 0.7689
Epoch 86/100
251/251 [==============================] - 0s 139us/step - loss: 0.6318 - accuracy: 0.7689
Epoch 87/100
251/251 [==============================] - 0s 139us/step - loss: 0.6377 - accuracy: 0.7729
Epoch 88/100
251/251 [==============================] - 0s 138us/step - loss:

[Parallel(n_jobs=1)]: Done  15 out of  15 | elapsed:   51.3s remaining:    0.0s


Epoch 1/10
250/250 [==============================] - 0s 667us/step - loss: 0.9042 - accuracy: 0.4600
Epoch 2/10
250/250 [==============================] - 0s 85us/step - loss: 0.7675 - accuracy: 0.7120
Epoch 3/10
250/250 [==============================] - 0s 85us/step - loss: 0.6912 - accuracy: 0.7280
Epoch 4/10
250/250 [==============================] - 0s 84us/step - loss: 0.6513 - accuracy: 0.7600
Epoch 5/10
250/250 [==============================] - 0s 86us/step - loss: 0.6402 - accuracy: 0.7680
Epoch 6/10
250/250 [==============================] - 0s 83us/step - loss: 0.6357 - accuracy: 0.7760
Epoch 7/10
250/250 [==============================] - 0s 84us/step - loss: 0.6277 - accuracy: 0.7880
Epoch 8/10
250/250 [==============================] - 0s 86us/step - loss: 0.6337 - accuracy: 0.7960
Epoch 9/10
250/250 [==============================] - 0s 84us/step - loss: 0.6352 - accuracy: 0.7760
Epoch 10/10
250/250 [==============================] - 0s 51us/step
[CV]  batch_size=20, e

[Parallel(n_jobs=1)]: Done  16 out of  16 | elapsed:   52.4s remaining:    0.0s


Epoch 1/10
250/250 [==============================] - 0s 768us/step - loss: 0.8810 - accuracy: 0.7600
Epoch 2/10
250/250 [==============================] - 0s 99us/step - loss: 0.7382 - accuracy: 0.7360
Epoch 3/10
250/250 [==============================] - 0s 111us/step - loss: 0.6594 - accuracy: 0.7560
Epoch 4/10
250/250 [==============================] - 0s 101us/step - loss: 0.6502 - accuracy: 0.7800
Epoch 5/10
250/250 [==============================] - 0s 106us/step - loss: 0.6517 - accuracy: 0.7960
Epoch 6/10
250/250 [==============================] - 0s 95us/step - loss: 0.6306 - accuracy: 0.7960
Epoch 7/10
250/250 [==============================] - 0s 84us/step - loss: 0.6338 - accuracy: 0.7960
Epoch 8/10
250/250 [==============================] - 0s 83us/step - loss: 0.6265 - accuracy: 0.8040
Epoch 9/10
250/250 [==============================] - 0s 83us/step - loss: 0.6258 - accuracy: 0.8040
Epoch 10/10
250/250 [==============================] - 0s 45us/step
[CV]  batch_size=20

[Parallel(n_jobs=1)]: Done  17 out of  17 | elapsed:   53.4s remaining:    0.0s


Epoch 1/10
250/250 [==============================] - 0s 661us/step - loss: 0.8933 - accuracy: 0.6800
Epoch 2/10
250/250 [==============================] - 0s 90us/step - loss: 0.7641 - accuracy: 0.7120
Epoch 3/10
250/250 [==============================] - 0s 85us/step - loss: 0.7075 - accuracy: 0.7400
Epoch 4/10
250/250 [==============================] - 0s 85us/step - loss: 0.6938 - accuracy: 0.7520
Epoch 5/10
250/250 [==============================] - 0s 87us/step - loss: 0.6859 - accuracy: 0.7600
Epoch 6/10
250/250 [==============================] - 0s 85us/step - loss: 0.6824 - accuracy: 0.7560
Epoch 7/10
250/250 [==============================] - 0s 85us/step - loss: 0.6765 - accuracy: 0.7680
Epoch 8/10
250/250 [==============================] - 0s 91us/step - loss: 0.6740 - accuracy: 0.7680
Epoch 9/10
250/250 [==============================] - 0s 83us/step - loss: 0.6768 - accuracy: 0.7720
Epoch 10/10
250/250 [==============================] - 0s 46us/step
[CV]  batch_size=20, e

[Parallel(n_jobs=1)]: Done  18 out of  18 | elapsed:   54.4s remaining:    0.0s


Epoch 1/10
251/251 [==============================] - 0s 660us/step - loss: 0.8950 - accuracy: 0.5538
Epoch 2/10
251/251 [==============================] - 0s 84us/step - loss: 0.7927 - accuracy: 0.7171
Epoch 3/10
251/251 [==============================] - 0s 82us/step - loss: 0.7027 - accuracy: 0.7092
Epoch 4/10
251/251 [==============================] - 0s 86us/step - loss: 0.6939 - accuracy: 0.7331
Epoch 5/10
251/251 [==============================] - 0s 89us/step - loss: 0.6843 - accuracy: 0.7530
Epoch 6/10
251/251 [==============================] - 0s 87us/step - loss: 0.6780 - accuracy: 0.7649
Epoch 7/10
251/251 [==============================] - 0s 86us/step - loss: 0.6729 - accuracy: 0.7689
Epoch 8/10
251/251 [==============================] - 0s 86us/step - loss: 0.6687 - accuracy: 0.7689
Epoch 9/10
251/251 [==============================] - 0s 85us/step - loss: 0.6687 - accuracy: 0.7649
Epoch 10/10
251/251 [==============================] - 0s 46us/step
[CV]  batch_size=20, e

[Parallel(n_jobs=1)]: Done  19 out of  19 | elapsed:   55.5s remaining:    0.0s


Epoch 1/10
251/251 [==============================] - 0s 656us/step - loss: 0.8676 - accuracy: 0.7331
Epoch 2/10
251/251 [==============================] - 0s 83us/step - loss: 0.7259 - accuracy: 0.7331
Epoch 3/10
251/251 [==============================] - 0s 82us/step - loss: 0.6697 - accuracy: 0.7371
Epoch 4/10
251/251 [==============================] - 0s 82us/step - loss: 0.6533 - accuracy: 0.7689
Epoch 5/10
251/251 [==============================] - 0s 83us/step - loss: 0.6415 - accuracy: 0.7530
Epoch 6/10
251/251 [==============================] - 0s 83us/step - loss: 0.6364 - accuracy: 0.7490
Epoch 7/10
251/251 [==============================] - 0s 80us/step - loss: 0.6344 - accuracy: 0.7490
Epoch 8/10
251/251 [==============================] - 0s 82us/step - loss: 0.6330 - accuracy: 0.7729
Epoch 9/10
251/251 [==============================] - 0s 83us/step - loss: 0.6311 - accuracy: 0.7649
Epoch 10/10
251/251 [==============================] - 0s 46us/step
[CV]  batch_size=20, e

250/250 [==============================] - 0s 85us/step - loss: 0.6244 - accuracy: 0.8040
Epoch 17/50
250/250 [==============================] - 0s 83us/step - loss: 0.6268 - accuracy: 0.8000
Epoch 18/50
250/250 [==============================] - 0s 79us/step - loss: 0.6273 - accuracy: 0.7960
Epoch 19/50
250/250 [==============================] - 0s 81us/step - loss: 0.6277 - accuracy: 0.8160
Epoch 20/50
250/250 [==============================] - 0s 79us/step - loss: 0.6271 - accuracy: 0.8080
Epoch 21/50
250/250 [==============================] - 0s 88us/step - loss: 0.6275 - accuracy: 0.8040
Epoch 22/50
250/250 [==============================] - 0s 86us/step - loss: 0.6260 - accuracy: 0.8040
Epoch 23/50
250/250 [==============================] - 0s 81us/step - loss: 0.6237 - accuracy: 0.8080
Epoch 24/50
250/250 [==============================] - 0s 88us/step - loss: 0.6247 - accuracy: 0.7960
Epoch 25/50
250/250 [==============================] - 0s 80us/step - loss: 0.6228 - accuracy:

Epoch 44/50
250/250 [==============================] - 0s 108us/step - loss: 0.6722 - accuracy: 0.7640
Epoch 45/50
250/250 [==============================] - 0s 84us/step - loss: 0.6742 - accuracy: 0.7600
Epoch 46/50
250/250 [==============================] - 0s 78us/step - loss: 0.6769 - accuracy: 0.7640
Epoch 47/50
250/250 [==============================] - 0s 86us/step - loss: 0.6835 - accuracy: 0.7560
Epoch 48/50
250/250 [==============================] - 0s 77us/step - loss: 0.6720 - accuracy: 0.7520
Epoch 49/50
250/250 [==============================] - 0s 78us/step - loss: 0.6725 - accuracy: 0.7600
Epoch 50/50
250/250 [==============================] - 0s 43us/step
[CV]  batch_size=20, epochs=50, score=(train=0.776, test=0.794), total=   1.8s
[CV] batch_size=20, epochs=50 ........................................
Epoch 1/50
251/251 [==============================] - 0s 680us/step - loss: 0.8881 - accuracy: 0.7131
Epoch 2/50
251/251 [==============================] - 0s 81us/step 

251/251 [==============================] - 0s 87us/step - loss: 0.6294 - accuracy: 0.7649
Epoch 20/50
251/251 [==============================] - 0s 83us/step - loss: 0.6314 - accuracy: 0.7649
Epoch 21/50
251/251 [==============================] - 0s 80us/step - loss: 0.6313 - accuracy: 0.7729
Epoch 22/50
251/251 [==============================] - 0s 80us/step - loss: 0.6300 - accuracy: 0.7689
Epoch 23/50
251/251 [==============================] - 0s 80us/step - loss: 0.6305 - accuracy: 0.7769
Epoch 24/50
251/251 [==============================] - 0s 79us/step - loss: 0.6289 - accuracy: 0.7729
Epoch 25/50
251/251 [==============================] - 0s 80us/step - loss: 0.6325 - accuracy: 0.7729
Epoch 26/50
251/251 [==============================] - 0s 86us/step - loss: 0.6324 - accuracy: 0.7689
Epoch 27/50
251/251 [==============================] - 0s 87us/step - loss: 0.6311 - accuracy: 0.7729
Epoch 28/50
251/251 [==============================] - 0s 89us/step - loss: 0.6305 - accuracy:

250/250 [==============================] - 0s 85us/step - loss: 0.6308 - accuracy: 0.7800
Epoch 47/100
250/250 [==============================] - 0s 83us/step - loss: 0.6281 - accuracy: 0.7800
Epoch 48/100
250/250 [==============================] - 0s 83us/step - loss: 0.6280 - accuracy: 0.7920
Epoch 49/100
250/250 [==============================] - 0s 80us/step - loss: 0.6272 - accuracy: 0.7920
Epoch 50/100
250/250 [==============================] - 0s 83us/step - loss: 0.6257 - accuracy: 0.7800
Epoch 51/100
250/250 [==============================] - 0s 81us/step - loss: 0.6272 - accuracy: 0.7800
Epoch 52/100
250/250 [==============================] - 0s 82us/step - loss: 0.6248 - accuracy: 0.7840
Epoch 53/100
250/250 [==============================] - 0s 81us/step - loss: 0.6250 - accuracy: 0.7720
Epoch 54/100
250/250 [==============================] - 0s 82us/step - loss: 0.6263 - accuracy: 0.7720
Epoch 55/100
250/250 [==============================] - 0s 81us/step - loss: 0.6241 - 

250/250 [==============================] - 0s 93us/step - loss: 0.6241 - accuracy: 0.8000
Epoch 24/100
250/250 [==============================] - 0s 79us/step - loss: 0.6233 - accuracy: 0.8040
Epoch 25/100
250/250 [==============================] - 0s 82us/step - loss: 0.6236 - accuracy: 0.8040
Epoch 26/100
250/250 [==============================] - 0s 86us/step - loss: 0.6240 - accuracy: 0.8080
Epoch 27/100
250/250 [==============================] - 0s 85us/step - loss: 0.6220 - accuracy: 0.8120
Epoch 28/100
250/250 [==============================] - 0s 79us/step - loss: 0.6243 - accuracy: 0.8120
Epoch 29/100
250/250 [==============================] - 0s 79us/step - loss: 0.6245 - accuracy: 0.8080
Epoch 30/100
250/250 [==============================] - 0s 167us/step - loss: 0.6232 - accuracy: 0.8040
Epoch 31/100
250/250 [==============================] - 0s 155us/step - loss: 0.6244 - accuracy: 0.8040
Epoch 32/100
250/250 [==============================] - 0s 110us/step - loss: 0.6239

Epoch 1/100
250/250 [==============================] - 0s 647us/step - loss: 0.9121 - accuracy: 0.3920
Epoch 2/100
250/250 [==============================] - 0s 87us/step - loss: 0.8189 - accuracy: 0.6920
Epoch 3/100
250/250 [==============================] - 0s 82us/step - loss: 0.7293 - accuracy: 0.7320
Epoch 4/100
250/250 [==============================] - 0s 82us/step - loss: 0.7204 - accuracy: 0.7160
Epoch 5/100
250/250 [==============================] - 0s 83us/step - loss: 0.7041 - accuracy: 0.7440
Epoch 6/100
250/250 [==============================] - 0s 82us/step - loss: 0.6908 - accuracy: 0.7440
Epoch 7/100
250/250 [==============================] - 0s 82us/step - loss: 0.6871 - accuracy: 0.7560
Epoch 8/100
250/250 [==============================] - 0s 83us/step - loss: 0.6828 - accuracy: 0.7520
Epoch 9/100
250/250 [==============================] - 0s 81us/step - loss: 0.6800 - accuracy: 0.7480
Epoch 10/100
250/250 [==============================] - 0s 79us/step - loss: 0.68

250/250 [==============================] - 0s 83us/step - loss: 0.6718 - accuracy: 0.7600
Epoch 81/100
250/250 [==============================] - 0s 81us/step - loss: 0.6723 - accuracy: 0.7720
Epoch 82/100
250/250 [==============================] - 0s 78us/step - loss: 0.6718 - accuracy: 0.7680
Epoch 83/100
250/250 [==============================] - 0s 77us/step - loss: 0.6725 - accuracy: 0.7640
Epoch 84/100
250/250 [==============================] - 0s 86us/step - loss: 0.6719 - accuracy: 0.7640
Epoch 85/100
250/250 [==============================] - 0s 86us/step - loss: 0.6760 - accuracy: 0.7560
Epoch 86/100
250/250 [==============================] - 0s 78us/step - loss: 0.6732 - accuracy: 0.7640
Epoch 87/100
250/250 [==============================] - 0s 85us/step - loss: 0.6733 - accuracy: 0.7680
Epoch 88/100
250/250 [==============================] - 0s 87us/step - loss: 0.6718 - accuracy: 0.7600
Epoch 89/100
250/250 [==============================] - 0s 86us/step - loss: 0.6716 - 

251/251 [==============================] - 0s 87us/step - loss: 0.6691 - accuracy: 0.7809
Epoch 58/100
251/251 [==============================] - 0s 85us/step - loss: 0.6650 - accuracy: 0.7729
Epoch 59/100
251/251 [==============================] - 0s 85us/step - loss: 0.6645 - accuracy: 0.7729
Epoch 60/100
251/251 [==============================] - 0s 87us/step - loss: 0.6659 - accuracy: 0.7610
Epoch 61/100
251/251 [==============================] - 0s 86us/step - loss: 0.6630 - accuracy: 0.7689
Epoch 62/100
251/251 [==============================] - 0s 88us/step - loss: 0.6633 - accuracy: 0.7769
Epoch 63/100
251/251 [==============================] - 0s 86us/step - loss: 0.6662 - accuracy: 0.7729
Epoch 64/100
251/251 [==============================] - 0s 88us/step - loss: 0.6669 - accuracy: 0.7849
Epoch 65/100
251/251 [==============================] - 0s 88us/step - loss: 0.6673 - accuracy: 0.7849
Epoch 66/100
251/251 [==============================] - 0s 87us/step - loss: 0.6640 - 

251/251 [==============================] - 0s 99us/step - loss: 0.6283 - accuracy: 0.7729
Epoch 34/100
251/251 [==============================] - 0s 91us/step - loss: 0.6294 - accuracy: 0.7689
Epoch 35/100
251/251 [==============================] - 0s 86us/step - loss: 0.6363 - accuracy: 0.7570
Epoch 36/100
251/251 [==============================] - 0s 108us/step - loss: 0.6278 - accuracy: 0.7610
Epoch 37/100
251/251 [==============================] - 0s 94us/step - loss: 0.6304 - accuracy: 0.7689
Epoch 38/100
251/251 [==============================] - 0s 102us/step - loss: 0.6341 - accuracy: 0.7610
Epoch 39/100
251/251 [==============================] - 0s 95us/step - loss: 0.6295 - accuracy: 0.7689
Epoch 40/100
251/251 [==============================] - 0s 99us/step - loss: 0.6300 - accuracy: 0.7729
Epoch 41/100
251/251 [==============================] - 0s 89us/step - loss: 0.6309 - accuracy: 0.7809
Epoch 42/100
251/251 [==============================] - 0s 85us/step - loss: 0.6300 

250/250 [==============================] - 0s 35us/step
[CV]  batch_size=40, epochs=10, score=(train=0.780, test=0.762), total=   0.9s
[CV] batch_size=40, epochs=10 ........................................
Epoch 1/10
250/250 [==============================] - 0s 658us/step - loss: 0.9137 - accuracy: 0.4920
Epoch 2/10
250/250 [==============================] - 0s 51us/step - loss: 0.8947 - accuracy: 0.7280
Epoch 3/10
250/250 [==============================] - 0s 54us/step - loss: 0.8451 - accuracy: 0.7320
Epoch 4/10
250/250 [==============================] - 0s 52us/step - loss: 0.7552 - accuracy: 0.7240
Epoch 5/10
250/250 [==============================] - 0s 52us/step - loss: 0.6844 - accuracy: 0.7240
Epoch 6/10
250/250 [==============================] - 0s 53us/step - loss: 0.6642 - accuracy: 0.7440
Epoch 7/10
250/250 [==============================] - 0s 52us/step - loss: 0.6597 - accuracy: 0.7640
Epoch 8/10
250/250 [==============================] - 0s 58us/step - loss: 0.6480 - ac

250/250 [==============================] - 0s 55us/step - loss: 0.6295 - accuracy: 0.7840
Epoch 29/50
250/250 [==============================] - 0s 60us/step - loss: 0.6286 - accuracy: 0.7880
Epoch 30/50
250/250 [==============================] - 0s 50us/step - loss: 0.6265 - accuracy: 0.7800
Epoch 31/50
250/250 [==============================] - 0s 51us/step - loss: 0.6256 - accuracy: 0.7840
Epoch 32/50
250/250 [==============================] - 0s 56us/step - loss: 0.6255 - accuracy: 0.7760
Epoch 33/50
250/250 [==============================] - 0s 71us/step - loss: 0.6249 - accuracy: 0.7760
Epoch 34/50
250/250 [==============================] - 0s 71us/step - loss: 0.6256 - accuracy: 0.7840
Epoch 35/50
250/250 [==============================] - 0s 81us/step - loss: 0.6252 - accuracy: 0.7880
Epoch 36/50
250/250 [==============================] - 0s 88us/step - loss: 0.6258 - accuracy: 0.7920
Epoch 37/50
250/250 [==============================] - 0s 64us/step - loss: 0.6239 - accuracy:

250/250 [==============================] - 0s 70us/step - loss: 0.7929 - accuracy: 0.7120
Epoch 4/50
250/250 [==============================] - 0s 66us/step - loss: 0.7263 - accuracy: 0.7240
Epoch 5/50
250/250 [==============================] - 0s 62us/step - loss: 0.7131 - accuracy: 0.7440
Epoch 6/50
250/250 [==============================] - 0s 59us/step - loss: 0.7025 - accuracy: 0.7520
Epoch 7/50
250/250 [==============================] - 0s 53us/step - loss: 0.6990 - accuracy: 0.7600
Epoch 8/50
250/250 [==============================] - 0s 54us/step - loss: 0.6887 - accuracy: 0.7560
Epoch 9/50
250/250 [==============================] - 0s 52us/step - loss: 0.6860 - accuracy: 0.7560
Epoch 10/50
250/250 [==============================] - 0s 52us/step - loss: 0.6834 - accuracy: 0.7520
Epoch 11/50
250/250 [==============================] - 0s 52us/step - loss: 0.6816 - accuracy: 0.7600
Epoch 12/50
250/250 [==============================] - 0s 53us/step - loss: 0.6788 - accuracy: 0.764

251/251 [==============================] - 0s 97us/step - loss: 0.6648 - accuracy: 0.7610
Epoch 32/50
251/251 [==============================] - 0s 76us/step - loss: 0.6649 - accuracy: 0.7649
Epoch 33/50
251/251 [==============================] - 0s 60us/step - loss: 0.6635 - accuracy: 0.7610
Epoch 34/50
251/251 [==============================] - 0s 53us/step - loss: 0.6638 - accuracy: 0.7610
Epoch 35/50
251/251 [==============================] - 0s 59us/step - loss: 0.6644 - accuracy: 0.7649
Epoch 36/50
251/251 [==============================] - 0s 52us/step - loss: 0.6631 - accuracy: 0.7570
Epoch 37/50
251/251 [==============================] - 0s 58us/step - loss: 0.6634 - accuracy: 0.7729
Epoch 38/50
251/251 [==============================] - 0s 65us/step - loss: 0.6641 - accuracy: 0.7610
Epoch 39/50
251/251 [==============================] - 0s 58us/step - loss: 0.6647 - accuracy: 0.7649
Epoch 40/50
251/251 [==============================] - 0s 54us/step - loss: 0.6642 - accuracy:

250/250 [==============================] - 0s 59us/step - loss: 0.6479 - accuracy: 0.7680
Epoch 7/100
250/250 [==============================] - 0s 57us/step - loss: 0.6409 - accuracy: 0.7800
Epoch 8/100
250/250 [==============================] - 0s 54us/step - loss: 0.6366 - accuracy: 0.7840
Epoch 9/100
250/250 [==============================] - 0s 55us/step - loss: 0.6306 - accuracy: 0.7760
Epoch 10/100
250/250 [==============================] - 0s 54us/step - loss: 0.6310 - accuracy: 0.7760
Epoch 11/100
250/250 [==============================] - 0s 55us/step - loss: 0.6304 - accuracy: 0.7760
Epoch 12/100
250/250 [==============================] - 0s 54us/step - loss: 0.6285 - accuracy: 0.7760
Epoch 13/100
250/250 [==============================] - 0s 55us/step - loss: 0.6282 - accuracy: 0.7720
Epoch 14/100
250/250 [==============================] - 0s 59us/step - loss: 0.6303 - accuracy: 0.7800
Epoch 15/100
250/250 [==============================] - 0s 62us/step - loss: 0.6281 - acc

250/250 [==============================] - 0s 68us/step - loss: 0.6251 - accuracy: 0.7760
Epoch 86/100
250/250 [==============================] - 0s 65us/step - loss: 0.6235 - accuracy: 0.7840
Epoch 87/100
250/250 [==============================] - 0s 61us/step - loss: 0.6250 - accuracy: 0.7840
Epoch 88/100
250/250 [==============================] - 0s 61us/step - loss: 0.6236 - accuracy: 0.7840
Epoch 89/100
250/250 [==============================] - 0s 54us/step - loss: 0.6237 - accuracy: 0.7880
Epoch 90/100
250/250 [==============================] - 0s 59us/step - loss: 0.6279 - accuracy: 0.7920
Epoch 91/100
250/250 [==============================] - 0s 51us/step - loss: 0.6253 - accuracy: 0.7880
Epoch 92/100
250/250 [==============================] - 0s 60us/step - loss: 0.6240 - accuracy: 0.7760
Epoch 93/100
250/250 [==============================] - 0s 53us/step - loss: 0.6324 - accuracy: 0.7800
Epoch 94/100
250/250 [==============================] - 0s 52us/step - loss: 0.6338 - 

250/250 [==============================] - 0s 75us/step - loss: 0.6229 - accuracy: 0.8120
Epoch 63/100
250/250 [==============================] - 0s 63us/step - loss: 0.6260 - accuracy: 0.8080
Epoch 64/100
250/250 [==============================] - 0s 56us/step - loss: 0.6252 - accuracy: 0.8120
Epoch 65/100
250/250 [==============================] - 0s 53us/step - loss: 0.6234 - accuracy: 0.8120
Epoch 66/100
250/250 [==============================] - 0s 52us/step - loss: 0.6231 - accuracy: 0.8040
Epoch 67/100
250/250 [==============================] - 0s 65us/step - loss: 0.6239 - accuracy: 0.8040
Epoch 68/100
250/250 [==============================] - 0s 69us/step - loss: 0.6235 - accuracy: 0.8040
Epoch 69/100
250/250 [==============================] - 0s 54us/step - loss: 0.6253 - accuracy: 0.8120
Epoch 70/100
250/250 [==============================] - 0s 53us/step - loss: 0.6229 - accuracy: 0.8040
Epoch 71/100
250/250 [==============================] - 0s 64us/step - loss: 0.6233 - 

250/250 [==============================] - 0s 90us/step - loss: 0.6718 - accuracy: 0.7640
Epoch 40/100
250/250 [==============================] - 0s 89us/step - loss: 0.6710 - accuracy: 0.7520
Epoch 41/100
250/250 [==============================] - 0s 78us/step - loss: 0.6753 - accuracy: 0.7600
Epoch 42/100
250/250 [==============================] - 0s 74us/step - loss: 0.6720 - accuracy: 0.7560
Epoch 43/100
250/250 [==============================] - 0s 77us/step - loss: 0.6693 - accuracy: 0.7560
Epoch 44/100
250/250 [==============================] - 0s 88us/step - loss: 0.6703 - accuracy: 0.7480
Epoch 45/100
250/250 [==============================] - 0s 88us/step - loss: 0.6687 - accuracy: 0.7640
Epoch 46/100
250/250 [==============================] - 0s 87us/step - loss: 0.6716 - accuracy: 0.7720
Epoch 47/100
250/250 [==============================] - 0s 85us/step - loss: 0.6705 - accuracy: 0.7760
Epoch 48/100
250/250 [==============================] - 0s 89us/step - loss: 0.6703 - 

251/251 [==============================] - 0s 62us/step - loss: 0.6632 - accuracy: 0.7769
Epoch 17/100
251/251 [==============================] - 0s 70us/step - loss: 0.6640 - accuracy: 0.7729
Epoch 18/100
251/251 [==============================] - 0s 57us/step - loss: 0.6668 - accuracy: 0.7689
Epoch 19/100
251/251 [==============================] - 0s 66us/step - loss: 0.6668 - accuracy: 0.7689
Epoch 20/100
251/251 [==============================] - 0s 65us/step - loss: 0.6652 - accuracy: 0.7649
Epoch 21/100
251/251 [==============================] - 0s 66us/step - loss: 0.6626 - accuracy: 0.7729
Epoch 22/100
251/251 [==============================] - 0s 73us/step - loss: 0.6621 - accuracy: 0.7809
Epoch 23/100
251/251 [==============================] - 0s 56us/step - loss: 0.6629 - accuracy: 0.7809
Epoch 24/100
251/251 [==============================] - 0s 79us/step - loss: 0.6629 - accuracy: 0.7729
Epoch 25/100
251/251 [==============================] - 0s 89us/step - loss: 0.6633 - 

251/251 [==============================] - 0s 68us/step - loss: 0.6629 - accuracy: 0.7689
Epoch 96/100
251/251 [==============================] - 0s 90us/step - loss: 0.6617 - accuracy: 0.7689
Epoch 97/100
251/251 [==============================] - 0s 98us/step - loss: 0.6650 - accuracy: 0.7729
Epoch 98/100
251/251 [==============================] - 0s 85us/step - loss: 0.6632 - accuracy: 0.7729
Epoch 99/100
251/251 [==============================] - 0s 77us/step - loss: 0.6635 - accuracy: 0.7729
Epoch 100/100
251/251 [==============================] - 0s 54us/step
[CV]  batch_size=40, epochs=100, score=(train=0.773, test=0.806), total=   3.0s
[CV] batch_size=40, epochs=100 .......................................
Epoch 1/100
251/251 [==============================] - 0s 683us/step - loss: 0.8987 - accuracy: 0.6614
Epoch 2/100
251/251 [==============================] - 0s 81us/step - loss: 0.8493 - accuracy: 0.7689
Epoch 3/100
251/251 [==============================] - 0s 63us/step - lo

251/251 [==============================] - 0s 57us/step - loss: 0.6296 - accuracy: 0.7729
Epoch 73/100
251/251 [==============================] - 0s 61us/step - loss: 0.6322 - accuracy: 0.7729
Epoch 74/100
251/251 [==============================] - 0s 54us/step - loss: 0.6298 - accuracy: 0.7769
Epoch 75/100
251/251 [==============================] - 0s 51us/step - loss: 0.6276 - accuracy: 0.7729
Epoch 76/100
251/251 [==============================] - 0s 60us/step - loss: 0.6272 - accuracy: 0.7769
Epoch 77/100
251/251 [==============================] - 0s 57us/step - loss: 0.6281 - accuracy: 0.7729
Epoch 78/100
251/251 [==============================] - 0s 57us/step - loss: 0.6299 - accuracy: 0.7649
Epoch 79/100
251/251 [==============================] - 0s 62us/step - loss: 0.6290 - accuracy: 0.7689
Epoch 80/100
251/251 [==============================] - 0s 65us/step - loss: 0.6285 - accuracy: 0.7769
Epoch 81/100
251/251 [==============================] - 0s 66us/step - loss: 0.6273 - 

[Parallel(n_jobs=1)]: Done  45 out of  45 | elapsed:  1.8min finished


Epoch 1/50
313/313 [==============================] - 0s 639us/step - loss: 0.8853 - accuracy: 0.7220
Epoch 2/50
313/313 [==============================] - 0s 63us/step - loss: 0.7756 - accuracy: 0.7316
Epoch 3/50
313/313 [==============================] - 0s 60us/step - loss: 0.6947 - accuracy: 0.7380
Epoch 4/50
313/313 [==============================] - 0s 51us/step - loss: 0.6826 - accuracy: 0.7508
Epoch 5/50
313/313 [==============================] - 0s 46us/step - loss: 0.6655 - accuracy: 0.7604
Epoch 6/50
313/313 [==============================] - 0s 46us/step - loss: 0.6574 - accuracy: 0.7636
Epoch 7/50
313/313 [==============================] - 0s 46us/step - loss: 0.6530 - accuracy: 0.7700
Epoch 8/50
313/313 [==============================] - 0s 46us/step - loss: 0.6494 - accuracy: 0.7796
Epoch 9/50
313/313 [==============================] - 0s 45us/step - loss: 0.6482 - accuracy: 0.7796
Epoch 10/50
313/313 [==============================] - 0s 48us/step - loss: 0.6459 - accur

In [14]:
def summarize_result(grid_results):
    print("Best: {0}, using {1}".format(grid_results.best_score_, grid_results.best_params_))
    print('\n')

    means_train = grid_results.cv_results_['mean_train_score']
    means_test = grid_results.cv_results_['mean_test_score']
    stds = grid_results.cv_results_['std_test_score']
    params = grid_results.cv_results_['params']

    for mean_train, mean_test, stdev, param in zip(means_train, means_test, stds, params):
        print('With: {0} -> mean training accuracy: {1}, mean & std of test accuracy: ({2}, {3})'.format(param, mean_train, mean_test, stdev))

summarize_result(grid_results)

Best: 0.7766001105308533, using {'batch_size': 40, 'epochs': 50}


With: {'batch_size': 10, 'epochs': 10} -> mean training accuracy: 0.7668079614639283, mean & std of test accuracy: (0.7541218757629394, 0.02844921519240059)
With: {'batch_size': 10, 'epochs': 50} -> mean training accuracy: 0.7779601693153382, mean & std of test accuracy: (0.767025101184845, 0.04881930392506942)
With: {'batch_size': 10, 'epochs': 100} -> mean training accuracy: 0.7763792753219605, mean & std of test accuracy: (0.7701997041702271, 0.047857557526636826)
With: {'batch_size': 20, 'epochs': 10} -> mean training accuracy: 0.7819665431976318, mean & std of test accuracy: (0.7669226884841919, 0.03754620986062043)
With: {'batch_size': 20, 'epochs': 50} -> mean training accuracy: 0.7851601719856263, mean & std of test accuracy: (0.767025101184845, 0.04881930392506942)
With: {'batch_size': 20, 'epochs': 100} -> mean training accuracy: 0.7827537775039672, mean & std of test accuracy: (0.770148491859436, 0.0413081858

In [15]:
from sklearn.metrics import classification_report, accuracy_score

y_pred = grid.predict(X_standardized)
print(accuracy_score(y_train, y_pred))
print(classification_report(y_train, y_pred))

X_test_standardized = scaler.transform(X_test)
y_pred = grid.predict(X_test_standardized)
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

313/313 [==============================] - 0s 49us/step
0.7667731629392971
              precision    recall  f1-score   support

           0       0.86      0.79      0.82       213
           1       0.62      0.72      0.66       100

    accuracy                           0.77       313
   macro avg       0.74      0.75      0.74       313
weighted avg       0.78      0.77      0.77       313

79/79 [==============================] - 0s 36us/step
0.8354430379746836
              precision    recall  f1-score   support

           0       0.88      0.86      0.87        49
           1       0.77      0.80      0.79        30

    accuracy                           0.84        79
   macro avg       0.82      0.83      0.83        79
weighted avg       0.84      0.84      0.84        79



Now we have results for the training and test accuracy scores. The test accuracy is even higher than the training accuracy, but we should not overvalue it, as the amount of test data is small. But the interesting observation is, the precision, recall and F1 scores for class 1 have increased. Again, I would not overvalue it, as the test set is small. But it is an interesting observation.